Trigger Refactor

slarticodefast opened the pull request:
## About the PR
This PR got soaped, but it was worth it.
It should make the trigger system much more powerful and flexible to use:

- Almost all triggers and trigger effects are now predicted
- All trigger effects can now either target the user or the triggered entity, selectable via a `bool` DataField.
- You can have multiple independent triggers on the same entity.
- Conditions like usedelays or whitelists were dehardcoded from some trigger effects, moved into separate components and can now be used with any trigger.

## Why / Balance
The trigger system is really powerful, but underused because it was missing flexibility and prediction. The refactor allows us to replace a lot of components that have been hardcoded for very specific purposes with small reusable building blocks.
For example
`ExtinguishOnInteract` would become `ExtinguishOnTrigger` and `TriggerOnInteract`, both of which could be combined with any other trigger.
Just to name a few other components that could be replaced with this:
`AnnounceOnSpawn`
`EmitSoundOnPickup`
`StaminaDamageOnEmbed`
`IgniteOnMeleeHit`
`UseDelayOnMeleeHit`
`DamageOnToolInteract`
`DamageOnInteract`
`SolutionInjectOnProjectileHit`
and a lot more.

Think of it as device linking in yaml. This should open up a lot of new possibilities for yaml contributors as well.

## Technical details
The PR is large in size, but the code is really simple.
I moved all trigger code from the Explosion namespace and other places into its own `Trigger` namespace, so that all existing trigger effects can be easily found. Since the triggers and trigger effects are very modular most of them are in their own system handling only a few subscriptions.
All code that has been moved to Shared should now be properly networked and predicted. A few exeptions are trigger effects that don't exist in shared, for example atmos and chat code.

Triggers and trigger effects now have a key string that is passed into the `Trigger` method and the `TriggerEvent`, and each effect will only be activated if the key is the same. This allows us to have multiple triggers on the same entity without them interfering with each other.

Conditions like Whitelists and UseDelays have been removed from all trigger effects, and are now instead checked in an `AttemptTriggerEvent`. This allows us to move them into their own components so they can be used with any trigger.

Each trigger effect now has the `TargetUser` datafield, allowing us to select if the effect will target the triggered entity or the user of the trigger. Before this was working different on a case by case basis and some trigger components were renamed to make this more consistent.

TODO before this can be merged:
- Do a lot of in-game testing to make sure all items are still working correctly.

## Media

https://github.com/user-attachments/assets/1808436d-e05f-47d3-8605-a7c20dc58374

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
All trigger related components and systems have been moved to the `Content.Shared.Client/Shared/Server.Trigger.*` namespace.

`SharedTriggerOnProximityComponent` was removed. Use `TriggerOnProximity` instead.
`TriggerOnTimedCollideComponent` now uses `TimeSpan`s instead of accumulating `frametime`.
`TriggerOnMobStateChangeComponent` no longer automatically starts timer triggers, this has to be explicitly linked via the same trigger key now.
`TriggerWhenEmptyComponent` has been renamed to `TriggerOnEmptyGunshotComponent`.
`TriggerImplantActionComponent` has been renamed to `TriggerOnActivateImplantComponent`.
`DamageUserOnTriggerComponent` has been removed in favor of `DamageOnTriggerComponent`. Set `TargetUser` to `true` to obtain the same result.
`GhostKickUserOnTriggerComponent` has been removed in favor of `GhostKickOnTriggerComponent`. Set `TargetUser` to `true` to obtain the same result.
`SoundOnTriggerComponent` has been removed because it was a duplicate of `EmitSoundOnTriggerComponent`.
The `CanAnchor` and `CanUnanchor` datafields in `AnchorOnTriggerComponent` now actually work (they did not do anything before).
`GibOnTriggerComponent` now targets the owning entity by default. Set `TargetUser` to `true` to keep the previous effect.
`ShockOnTriggerComponent` now has a `TargetContainer` bool for when it's used for equipment. Otherwise it will shock the owning entity.
`ShockOnTriggerComponent` no longer has the `Cooldown` datafield. Use `UseDelayOnTriggerComponent` and `UseDelayTriggerConditionComponent` instead.
`IgniteOnTriggerComponent` no longer plays a sound, use `EmitSoundOnTrigger instead`.
`PolymorphOnTriggerComponent` now targets the owning entity by default. Set `TargetUser` to `true` to keep the previous effect.
`RattleComponent` has been renamed to `RattleOnTrigger` and targets the owning entity by default. Set `TargetUser` to `true`  to keep the previous effect.
`MouseTrapComponent` now uses `ItemToggleComponent` for arming and dearming.
`MouseTrapSystem` has been moved to `Content.Shared.Mousetrap`.
`ActiveListenerComponent` has been moved to `Content.Shared.Speech`.
`OnUseTimerTrigger` has been split up into multiple components. Use `TimerTrigger` instead.
`SignallerComponent` no longer triggers when receiving a signal. Add `TriggerOnSignalComponent` to keep this effect.
`TimerTrigger` does not automatically activate when the item is used in hand, stuck on a wall or when receiving a device link signal. Use `TriggerOnUse`, `TriggerOnStuck`, or `TriggerOnSignal` to send a trigger to start the timer.
`ActiveTimerTrigger` is now a pure marker component without duplicating the datafields in `OnUseTimerTrigger`.
`UseVerbInstead` has been removed from `TimerTrigger`. Use `TriggerOnVerb` instead to start the timer.
`AutomatedTimerComponent` has been removed, use other triggers to start the timer the way you need.

`TriggerWhitelistComponent` has been renamed to `WhitelistTriggerConditionComponent`.
All whitelist checks have been removed from trigger components. Add `WhitelistTriggerConditionComponent` to obtain the same result.
All usedelay checks have been removed from trigger components. Use `UseDelayOnTriggerComponent` and `UseDelayTriggerConditionComponent` instead.

**Changelog**
no player facing changes

ScarKy0 commented:
You bought me at "endless admeme potential"

perryprog reviewed (commented):
Hope you don't mind the nitpicks. I personally appreciate getting them, so.

Haven't started testing much yet and I'm scared to look at much less review the yaml changes.

slarticodefast commented on Content.Shared/Trigger/TriggerEvent.cs:
```diff
@@ -0,0 +1,20 @@
+/// Raised whenever something is Triggered on the entity.
+/// </summary>
+[ByRefEvent]
+public record struct TriggerEvent(EntityUid Triggered, EntityUid? User = null, string? Key = null, bool Handled = false);
```
good catch, originally triggers were a broadcast event, so this was needed, but I decided to not keep that anymore because I want to split that off into its own component, something like `BroadcastOnTrigger`

slarticodefast commented on Content.Shared/Trigger/Systems/TriggerSystem.Timer.cs:
```diff
@@ -0,0 +1,179 @@
+
+        // This is somewhat inefficient, but its good enough. This is run rarely, and the lists should be short.
+
+        ent.Comp.DelayOptions.Sort();
```
Eh, that was like that before, I just moved it here. I'll check if I can clean it up a little.

slarticodefast commented on Content.Shared/Trigger/Components/Triggers/BaseTriggerOnXComponent.cs:
```diff
@@ -0,0 +1,17 @@
+    /// The key that the trigger will activate.
+    /// null will activate all triggers.
+    /// </summary>
+    [DataField, AutoNetworkedField]
```
The autonetworking still needs to be double checked everywhere, and the keys aren't even networked at the moment because the source generator only works on the direct class members, ignoring the members from the abstract parents.
I wanted the keys networked so that admins can VV them, and because I was thinking about adding API functions for dynamically linking triggers in the future, which is why I used inheritance for this (an interface would be the alternative). Think of it like device linking.

perryprog replied:
Yeah, realized it being VVable is nice ~~since not everyone knows about the magic that is loadprototype~~. And right, good point regarding source generators.

slarticodefast commented on Content.Shared/Trigger/Systems/TriggerOnGunEmptySystem.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<TriggerOnGunEmptyComponent, OnEmptyGunShotEvent>(OnEmptyGunShot);
```
Yeah, I was thinking about that. Sloth said systems have almost no computational overhead, so I was preferring to keep them apart from each other to not have merge conflicts whenever someone adds a new trigger and to not have 100 totally unrelated imports in the same system.

perryprog replied:
Wait true merge conflicts are real; this way better.

(Still recovering from a recent 700 commit upstream merge I did.)

perryprog commented on Content.Shared/Trigger/Components/TimerTriggerComponent.cs:
```diff
@@ -0,0 +1,99 @@
+    /// The time between beeps.
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public TimeSpan BeepInterval;
```
This should have a default of one second or something.
```suggestion
    public TimeSpan BeepInterval = TimeSpan.FromSeconds(1);
```

perryprog commented on Content.Shared/Trigger/Systems/RepulseAttractOnTriggerSystem.cs:
```diff
@@ -0,0 +1,34 @@
+            return;
+
+        var position = _transform.GetMapCoordinates(target.Value);
+        _repulse.TryRepulseAttract(position, args.User, ent.Comp.Speed, ent.Comp.Range, ent.Comp.Whitelist, ent.Comp.CollisionMask);
```
Not sure why but this doesn't seem to be getting predicted—the client hits this fine, it just doesn't do anything. I thought this portion of the throwing system worked fine with prediction but maybe not...

This can probably be a different PR though.

perryprog commented on Resources/Prototypes/Entities/Objects/Misc/subdermal_implants.yml:
```diff
@@ -27,6 +27,7 @@
       mobState:
       - Dead
     - type: EmitSoundOnTrigger
+      predicted: true
```
```suggestion
      predicted: false
```
Seems you can't predict here since this goes through an implant relay, which other clients won't know about.

perryprog replied:
Wait you use PlayPredicted not PlayLocal so uhhhh I don't know why this fixes this (yet)

perryprog commented on Content.Shared/Trigger/Systems/GibOnTriggerSystem.cs:
```diff
@@ -0,0 +1,40 @@
+                PredictedQueueDel(item);
+            }
+        }
+        _body.GibBody(ent.Owner, true);
```
Remembered another deleted review comment:
```suggestion
        _body.GibBody(target.Value, true);
```

deltanedas commented on Content.Shared/Trigger/Components/Effects/AlertLevelChangeOnTriggerComponent.cs:
```diff
@@ -0,0 +1,33 @@
+    /// The alert level to change to when triggered.
+    ///</summary>
+    [DataField, AutoNetworkedField]
+    public string Level = "blue";
```
AlertLevelPrototype is a thing though

perryprog replied:
Its name is a prank; it should be AlertsLevelPrototype. There's no way to have a ProtoId or similar that refers to a single alert level.

deltanedas commented on Content.Shared/Trigger/Components/Effects/AnchorOnTriggerComponent.cs:
```diff
@@ -0,0 +1,29 @@
+    /// Removes this component when triggered so it can only be activated once.
+    /// </summary>
+    [DataField, AutoNetworkedField]
+    public bool RemoveOnTrigger = true;
```
you can have add/remove compregistry on trigger to do this in a generic way

slarticodefast replied:
Will do in a future PR, for now I'm keeping this as it was.

slarticodefast commented on Content.Shared/HotPotato/SharedHotPotatoSystem.cs:
```diff
@@ -1,18 +1,69 @@
 
 namespace Content.Shared.HotPotato;
 
 public abstract class SharedHotPotatoSystem : EntitySystem
```
fixed anyways

slarticodefast commented on Content.Shared/Trigger/Components/Effects/EmitSoundOnTriggerComponent.cs:
```diff
@@ -0,0 +1,31 @@
+    [DataField(required: true), AutoNetworkedField]
+    public SoundSpecifier? Sound;
+
+    /// <summary>
+    /// Play the sound at the position instead of parented to the source entity.
+    /// Useful if the entity is deleted after.
+    /// </summary>
+    [DataField]
+    public bool Positional;
```
It's named like that in all the other emit sound components, so I would keep it consistent.

slarticodefast commented on Content.Shared/Trigger/Components/Effects/EmitSoundOnTriggerComponent.cs:
```diff
@@ -0,0 +1,31 @@
+    /// The <see cref="SoundSpecifier"/> to play.
+    /// </summary>
+    [DataField(required: true), AutoNetworkedField]
+    public SoundSpecifier? Sound;
```
Nullable for admeme VV purposes, otherwise it might complain if it's missing.
Also BaseEmitSound does the same.

slarticodefast commented on Content.Shared/Trigger/Components/Triggers/TriggerOnProximityComponent.cs:
```diff
@@ -0,0 +1,91 @@
+    /// What layer is the trigger fixture on?
+    /// </summary>
+    [DataField(customTypeSerializer: typeof(FlagSerializer<CollisionLayer>))]
+    public int Layer = (int)(CollisionGroup.MidImpassable | CollisionGroup.LowImpassable | CollisionGroup.HighImpassable);
```
probably, but it was like this before, so I'm not touching it

slarticodefast commented:
Thanks for the review! I addressed some of the points, will do the rest in the coming days.
Also some bugs I found while testing and still have to fix:
- C4 beeps rapidly when activated.
- C4 shows an error before exploding. The generic visualizer has a wrong rsi state, I just have to find the right one.

perryprog commented:
The rapid beeping was probably from the beep interval field having a default or zero seconds. For the sprite issue, one way you can workaround it is setting the appearance setdata call to be server only—it seems like it’s an issue with prediction timing, and the client not knowing that explosions mean the object gets deleted. (The real fix is probably making ExplodeOnTrigger do a predicted delete for the client or something.)

perryprog commented on Resources/Prototypes/Entities/Objects/Weapons/Bombs/plastic.yml:
```diff
@@ -33,7 +33,7 @@
         base:
           Primed: { state: primed }
-          Unprimed: { state: complete }
+          Unprimed: { state: icon }
```
So the sprite issue is an issue with most non-reusable bombs—I remember hot potato also having the issue, and I suspect there were others. I’m not sure if this is the best fix because of that. (It might also be an instance of bad yaml copy and pasting and it shouldn’t have this visualizer key at all? I’m not sure.)

slarticodefast replied:
Yeah, it's a case of bad copy paste. I blame the generic visualizer not validating the rsi states.
I'll fix all the other cases.

perryprog commented on Resources/Prototypes/Entities/Objects/Specific/Janitorial/soap.yml:
```diff
@@ -180,6 +180,7 @@
         - ItemMask
   - type: DeleteOnTrigger
   - type: EmitSoundOnTrigger
+    predicted: true
```
So this is an existing bug, so maybe just TODO this (because it's kinda a Balance Change™), but this needs a TriggerOnStep to actually work. It'll also need its audio to be set as positional if the DeleteOnTrigger stays. Additionally, this volume makes the splat effect literally inaudible for me. Maybe -8 or something would be better.

(The volume issue might actually be due to an engine issue I have yet to properly document, but it seems like different OpenAL implementations might not have consistent implementations of AL_GAIN which is... absurd. I'd be curious to know if the sound is audible for you as-is.)

perryprog reviewed (commented):
You didn't touch this file so I can't put the comment here, but for FlashPayload there's a bug I noticed that already existed, but it's an easy enough fix. The flash payload needs its EmitAudioOnTrigger to be set as positional as the payload is also deleted on trigger, causing you to never hear the sound it emits. (Arguably, the sound should just be parented to the modular grenade itself, but it's just a one second sound effect so I don't think it matters.)

I'm looking through over EmitSoundOnTrigger usages and noticing a lot that could also be tagged predicted, but this is pretty low priority of course (BasePlushie, toy horns, skub, everything in toys.yml, desk bell are the ones I see).

perryprog commented:
Hmm. It's not much better, but what about a ToggleComp field for only showing the popups in response to user input? That probably requires just as much argument-threading, though.

Alternatively a PopupOnTrigger that has a different key in and is fired by TriggerOnUse?

>I would have expected mob state changes to be predicted, but apparently the event isn't raised on the client at all.

It seems like mob state changes are networked—it's just that despite ImplantedComponent being a networked component, its system is exclusively serverside and is where ImplantContainer gets instantiated, so other clients aren't networked a given person's implants (probably fine because cheat clients™), meaning the mob state relay can't be predicted.

perryprog reviewed (approved):
Now your challenge is to get an actual maint to review :)

beck-thompson commented:
> Now your challenge is to get an actual maint to review :)

you basically count as a maint review imo :laughing:

slarticodefast commented on Content.Server/Defusable/Systems/DefusableSystem.cs:
```diff
@@ -139,16 +139,9 @@ public void TryStartCountdown(EntityUid uid, EntityUid user, DefusableComponent
 
         _popup.PopupEntity(Loc.GetString("defusable-popup-begun", ("name", uid)), uid);
-        if (TryComp<OnUseTimerTriggerComponent>(uid, out var timerTrigger))
+        if (TryComp<TimerTriggerComponent>(uid, out var timerTrigger))
```
The `Resolve` logs an error when you try to activate a timer on an entity that doesn't have the component.
Apparently the defusable system was made with bombs in mind that don't have timers triggers, since other effects could happen in `BombArmedEvent` subbscriptions.

beck-thompson replied:
ok bruh well its fine then!

slarticodefast commented on Content.Server/Nutrition/EntitySystems/CreamPieSystem.cs:
```diff
@@ -77,15 +77,9 @@ private void ActivatePayload(EntityUid uid)
                 if (_itemSlots.TryEject(uid, itemSlot, user: null, out var item))
                 {
-                    if (TryComp<OnUseTimerTriggerComponent>(item.Value, out var timerTrigger))
+                    if (TryComp<TimerTriggerComponent>(item.Value, out var timerTrigger))
```
same thing

slarticodefast commented on Content.Client/HotPotato/HotPotatoSystem.cs:
```diff
@@ -10,6 +10,7 @@ public sealed class HotPotatoSystem : SharedHotPotatoSystem
     [Dependency] private readonly IRobustRandom _random = default!;
     [Dependency] private readonly SharedTransformSystem _transform = default!;
 
+    // TODO: particle system
```
I turned the hardcoded string into an EntProtoId at least.

> there is a lot of broken stuff 

What exactly is broken besides the fact that it should be replaced with particles?

perryprog replied:
If I recall correctly it's currently throwing client-side errors on the person performing the melee hit to transfer it.

slarticodefast replied:
I couldn't reproduce any errors, but I fixed the popups.

beck-thompson reviewed (approved):
Looks good! Putting do not merge because I think scar wants to take a look and we should wait until stable is finished when we merge this. As soon as that's ready though lets get a full 2 weeks of testing 🥳 (I think it should be pretty bug free!)

ScarKy0 reviewed (approved):
The most admin abusable PR ever and its not even my birthday.

ScarKy0 commented:
<img width="712" height="530" alt="image" src="https://github.com/user-attachments/assets/99a1f417-905f-414d-8eed-61c1b8d2d8a9" />

perryprog commented:
oof ouchie my merge conflicts
